src/lib/png/png.c

```diff
@@ -0,0 +1,976 @@
+		case 1:
+			printf("TODO : support of pallete images didn't add.\n");
+			return -1;
+		case 2:
+			png_head->color_type = PNG_TRCL;
+			break;
+
+		case 4:
+			png_head->color_type = PNG_TRCLA;
+			break;
+
+		case 6:
+			png_head->color_type = PNG_TRCLA;
+			break;
+
+		default:
+			printf("Error : invalid color type.\n");
+			return -1;
+	}
+
+	/* read compression method, put in header and check */
+	if (read(fd, &(png_head->compression_method), sizeof(png_head->compression_method))
+												!= sizeof(png_head->compression_method)) {
+		printf("Error : can't read IHDR chunk.\n");
+		return -1;
+	}
+	if (png_head->compression_method != 0) {
+		printf("Error : invalid compression method.\n");
+		return -1;
+	}
+
+	/* read filter method, put in header and check */
+	if (read(fd, &(png_head->filter_method), sizeof(png_head->filter_method))
+											!= sizeof(png_head->filter_method)) {
+		printf("Error : can't read IHDR chunk.\n");
+		return -1;
+	}
+	if (png_head->filter_method != 0) {
+		printf("Error : invalid filter method.\n");
+		return -1;
+	}
+
+	/* read interlace method, put in header and check */
+	if (read(fd, &(png_head->interlace_method), sizeof(png_head->interlace_method))
+												!= sizeof(png_head->interlace_method)) {
+		printf("Error : can't read IHDR chunk.\n");
+		return -1;
+	}
+	if (png_head->interlace_method != 0 && png_head->interlace_method != 1) {
+		printf("Error : invalid interlace method\n");
+		return -1;
+	}
+
+	if (read(fd, &crc, sizeof(crc)) != sizeof(crc)) {
+		printf("Error : can't read crc of IHDR chunk.\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+/* Now lib support main chunks and this fun is stub for work with other chunks*/
+static void read_chunk(int fd, uint8_t *chunk_name) {
+	return;
+}
+
+static int png_read(int fd, struct png *png_data) {
+	struct png_header png_head;
+	uint32_t chunk_length;
+	uint8_t chunk_name[4];
+	int pos = IHDR_END;
+
+	/**
+	 * IDAT_flag
+	 * 1 - IDAT was successfull read and decoded
+	 * 0 - IDAT not yet met
+	 */
+	int flag_IDAT = 0;
+
+	/* read IHDR chunk and put file info in struct png_header */
+	if (read(fd, &chunk_length, sizeof(chunk_length)) != sizeof(chunk_length)) {
+		printf("Error : can't read IHDR chunk.\n");
+		return -1;
+	}
+
+	if ((int)htonl(chunk_length) != IHDR_SIZE) {
+		printf("Error : invalid length of IHDR data.\n");
+		return -1;
+	}
+
+	if (read(fd, chunk_name, sizeof(chunk_name)) != sizeof(chunk_name)) {
+		printf("Error : can't read IHDR chunk.\n");
+		return -1;
+	}
+
+	if (memcmp(chunk_name, "IHDR", sizeof(chunk_name)) != 0) {
+		printf("Error : invalid IHDR chunk.\n");
+		return -1;
+	}
+
+	if (read_IHDR(fd, &png_head) != 0) {
+		return -1;
+	}
+
+	png_data->width = png_head.width;
+	png_data->height = png_head.height;
+	png_data->color = png_head.color_type;
+	png_data->image_size = png_head.width * png_head.height * png_head.color_type;
+	/* malloc mem for image */
+	png_data->image = (uint8_t*)malloc(png_data->image_size);
```

Comment:
Please, add check for successful memory allocation

Reply:
done in all places